src/crimson/os/seastore/random_block_manager/rbm_device.cc

```diff
@@ -19,40 +19,89 @@ namespace crimson::os::seastore::random_block_device {
 SET_SUBSYS(seastore_device);
 
 RBMDevice::mkfs_ret RBMDevice::do_mkfs(device_config_t config) {
-  LOG_PREFIX(RBMDevice::mkfs);
+  assert(shard_devices.local());
+  return shard_devices.local()->do_primary_mkfs(config
+  ).safe_then([this] {
+    return shard_devices.invoke_on_all([](auto &local_device) {
+      assert(local_device);
+      return local_device->do_shard_mkfs(
+      ).handle_error(
+	crimson::ct_error::assert_all{
+	  "Invalid error in RBMDevice::do_mkfs"
+      });
+    });
+  });
+}
+
+RBMDevice::mkfs_ret RBMDevice::do_primary_mkfs(device_config_t config) {
+  LOG_PREFIX(RBMDevice::do_primary_mkfs);
   return stat_device(
   ).handle_error(
     mkfs_ertr::pass_further{},
     crimson::ct_error::assert_all{
-    "Invalid error stat_device in RBMDevice::mkfs"}
+    "Invalid error stat_device in RBMDevice::do_primary_mkfs"}
   ).safe_then([this, FNAME, config=std::move(config)](auto st) {
     super.block_size = st.block_size;
     super.size = st.size;
     super.feature |= RBM_BITMAP_BLOCK_CRC;
     super.config = std::move(config);
     assert(super.journal_size);
     assert(super.size >= super.journal_size);
+
+    std::vector<rbm_shard_info_t> shard_infos(seastar::smp::count);
+    for (unsigned int i = 0; i < seastar::smp::count; i++) {
+      shard_infos[i].size = super.size / seastar::smp::count;
```

Comment:
super.size is the device total size, right? so some header and meta data also store within this size. could you give the layout of disk for superblock, header or metadata and each shard?

Reply:
Right. So, the layout is:

| super (4096) | journal head | journal | data |  -- | super (4096) | journal head | journal | data | ..
------------primary shrad--------------------        ----------------------shard-------------------

The primary super contains all shard information.

Reply:
seems only primary shard fill super, the other shards don't fill super. why keep the super block for other shard?

Reply:
I'm thinking of a backup super block like other filesystems. This might help to check inconsistency and recovery. So, this part is reserved for now. Also, this can be used as the extension of the existing primary's superblock if the superblock size exceeds the block size. If we want to save the amount of space (super block size x #shard number) by removing a super block in other shards, this is also fine with me for now.  But, we need to change the disk layout entirely if that blocks are needed.

Reply:
we can leave it reserved now.